File: lt0337.java
import java.util.HashMap;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x){
        val = x;
    }
};

class Solution {
    public int rob(TreeNode root) {
        if(root == null)
        return 0;

        if(record.get(root)!=null)
            return record.get(root);
        
        Integer noRobVal = rob(root.left) + rob(root.right);
        Integer RobVal = root.val;
        if(root.left !=null)
            RobVal = RobVal + rob(root.left.left) + rob(root.left.right);
        
        if(root.right !=null)
            RobVal = RobVal + rob(root.right.left) + rob(root.right.right);

        record.put(root, Math.max(noRobVal, RobVal));
        return record.get(root);
    }

    private HashMap<TreeNode, Integer> record = new HashMap<>();
}

File: code/lt0337.cpp
#include <algorithm>
#include <vector>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x):val(x), left(nullptr), right(nullptr){}
};

#if defined S1
//2019-07-23: 超时, 解法没问题
class Solution {
public:
    int rob(TreeNode* root) {
        return rob(root, false) ;
    }

    int rob(TreeNode* root, bool last_rob)
    {
        if(root == nullptr)
            return 0;
        
        int both_not_rob = rob(root->left,  false) + rob(root->right,  false);
        if(last_rob)
            return rob(root->left,  false) + rob(root->right,  false);
        else
            return max(root->val + rob(root->left,  true) + rob(root->right,  true), both_not_rob) ; 

    }
};

#else

#endif